pkg/controller/grafanadashboard/dashboard_controller.go

```diff
@@ -132,6 +134,57 @@ func (r *ReconcileGrafanaDashboard) Reconcile(request reconcile.Request) (reconc
 		return r.deleteDashboard(cr)
 	}
 
+	changed, hash := r.hasDashboardChanged(cr.Spec.Json, cr)
+	if changed {
+		cr.Status.LastConfig = hash
+		r.client.Update(context.TODO(), cr)
+	}
+
+	known, err := r.helper.IsKnownDashboard(cr)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	jsonTest, err := isJSON(cr.Spec.Json)
+	// Dashboard is in cm and JSON is valid
+	if known && jsonTest == true {
+		if !changed {
+			log.Info(fmt.Sprintf("dashboard '%s' reconciled but no changes", cr.Name))
+			return reconcile.Result{}, nil
+		}
+	}
+
+	// Dashboard is in cm and JSON is invalid
+	if jsonTest == false && changed {
+		jsonError := err.Error()
+		msg := fmt.Sprintf("Invalid JSON, Error: %s", jsonError)
+		common.AppendMessage(msg, cr)
+		if known {
+			log.Info(fmt.Sprintf("invalid JSON found in reconciled dashboard '%s'", cr.Name))
+			r.deleteDashboard(cr)
+			cr.Status.Phase = common.StatusResourceUninitialized
+			r.client.Update(context.TODO(), cr)
+			return reconcile.Result{}, nil
+		}
+		log.Info(fmt.Sprintf("invalid JSON found in dashboard '%s'", cr.Name))
+		r.client.Update(context.TODO(), cr)
+		return reconcile.Result{}, nil
+	}
+
+	if jsonTest == false && !changed {
+		if !known {
+			log.Info(fmt.Sprintf("dashboard with invalid JSON '%s' reconciled but no changes", cr.Name))
+			return reconcile.Result{}, nil
+		}
+		if known {
+			r.deleteDashboard(cr)
+			cr.Status.Phase = common.StatusResourceUninitialized
```

Comment:
Doesn't this lead to the dashboard being reimported?

Reply:
No, because we return reconcile.Result after it and the json check won't allow it to get imported the next time.

This is to catch where the json was valid, so the dashboard is in the configmap, but now the json has changed and is invalid.